### PR TITLE
Refactor: Make backend's outbuffer a little bit safer

### PR DESCRIPTION
**Pull request by Geod24**
Better reviewed commit-by-commit.

The first commit is the largest one, but it's also a mechanical change from "size" to "length" to bring the naming in line with the D convention, where "length" is the used length, while "capacity" is the total size of the buffer (not exposed here).

This is a range of commit made while working on various things in the backend (DWARF, ABI issues), hence the disconnect for some commit. However each commit should be self-explanatory (e.g. the DWARF documentation one).

### CODE REVIEW

File: src/dmd/backend/mscoffobj.d [hunk 1]
```diff
@@ -2077,16 +2077,12 @@ static if (0)
     int save = cast(int)buf.size();
     //dbg_printf("MsCoffObj_bytes(seg=%d, offset=x%lx, nbytes=%d, p=x%x)\n",
             //seg,offset,nbytes,p);
-    buf.setsize(cast(uint)offset);
-    buf.reserve(nbytes);
+    buf.position(cast(size_t)offset, nbytes);
```
Comment (thewilsonator) [inline src/dmd/backend/mscoffobj.d]: Not that It particularly matters or this PR, but this whole function `MsCoffObj_bytes ` appears to be an exact duplicate of `Obj_bytes` in both `machobj.d` and `elfobj.d`

Comment (Geod24) [reply src/dmd/backend/mscoffobj.d]: Yep! The backend has a lot of duplications like this... I'm trying to get rid of *some* of it to make my issues easier to fix.

**Comment by thewilsonator** [comment]
There seems to be a lot of errors that looks like 
```
dmd/backend/elfobj.d(547): Error: function `dmd.backend.global.size(uint)` is not callable using argument types `(Outbuffer*)`
dmd/backend/elfobj.d(547):        cannot pass argument `section_names` of type `Outbuffer*` to parameter `uint`
```

**Review by thewilsonator — APPROVED** [review]
Other than the failing to compile this looks fine to me.

**Comment by Geod24** [comment]
> There seems to be a lot of errors that looks like

Yep, I could only compile on Mac. Should be green now.

There were a couple of changes:
- Reworked `outbuf` destructor in the second commit;
- Reverted the change from `writeString` to `write`, instead introduced 2 `writeString` overloads that write a NUL-terminator (affected the DWARF switch to slices);
- A couple more place where `size` => `length` was missing

**Comment by benjones** [comment]
Should the slicing operations with strlen use root.string.toDString instead?

**Comment by Geod24** [comment]
> Should the slicing operations with strlen use root.string.toDString instead?

Ideally yes. However I had some issues when adding `root` import in the backend because `betterC` (e.g. importing array didn't work) so decided to keep it simple for now.

**Pull request merged**